modin/backends/pandas/query_compiler.py

```diff
@@ -561,21 +561,92 @@ def reset_index(self, **kwargs):
-                else "index"
-                if "index" not in self.columns
-                else "level_0"
-            )
-            new_self = self.insert(0, new_column_name, self.index)
+            # Some parts of index remain in place
+            if self.has_multiindex():
+                col_level = kwargs.get("col_level", 0)
+                col_fill = kwargs.get("col_fill", "")
+
+                # Convert multiindex into a dataframe with columns containing multiindex levels
+                mi_frame = self.index.to_frame()
+                # Overwrite constructed columns index because it ruins tuples with NaNs since
+                # tupleize_cols is True by default
+                mi_frame.columns = pandas.Index(self.index.names, tupleize_cols=False)
+
+                if level is not None:
+                    mi_frame = mi_frame.iloc[:, uniq_sorted_level]
+                if isinstance(self.columns, pandas.MultiIndex):
+                    items = []
+                    if col_fill is None:
+                        # Initialize col_fill if it is None.
+                        # This is some weird undocumented Pandas behavior to take first
+                        # element of the last column name.
+                        last_col_name = mi_frame.columns[-1]
+                        last_col_name = (
+                            list(last_col_name)
+                            if isinstance(last_col_name, tuple)
+                            else [last_col_name]
+                        )
+                        if len(last_col_name) not in (1, self.columns.nlevels):
+                            raise ValueError(
+                                "col_fill=None is incompatible "
+                                f"with incomplete column name {last_col_name}"
+                            )
+                        col_fill = last_col_name[0]
+                    for mi_frame_level in mi_frame.columns:
+                        # Expand tuples into separate items and fill the rest with col_fill
+                        item = tuple(
+                            [col_fill for i in range(0, col_level)]
+                            + (
+                                list(mi_frame_level)
+                                if isinstance(mi_frame_level, tuple)
+                                else [mi_frame_level]
+                            )
+                            + [
+                                col_fill
+                                for i in range(
+                                    col_level
+                                    + (
+                                        len(mi_frame_level)
+                                        if isinstance(mi_frame_level, tuple)
+                                        else 1
+                                    ),
+                                    self.columns.nlevels,
+                                )
+                            ]
+                        )
```

Comment:
```suggestion
                        top_level = [col_fill] * col_level
                        middle_level = (
                            list(mi_frame_level)
                            if isinstance(mi_frame_level, tuple)
                            else [mi_frame_level]
                        )
                        bottom_level = [col_fill] * (
                            self.columns.nlevels - (col_level + len(middle_level))
                        )
                        item = tuple(top_level + middle_level + bottom_level)
```
This `item` construction looks heavy to read, can we change it somehow?

BTW, why do we need to convert `mi_frame_level` to the list when it's tuple?

Reply:
Ok this looks like a good suggestion. And `mi_frame_level` has to be converted into a list for concatenation with `[col_fill]` lists before and after. Tuples cannot be concatenated.